src/main/java/com/monumental/services/MonumentService.java

```diff
@@ -22,5 +32,72 @@ public MonumentService() {
 
     }
 
-    // Add Monument specific database operations here if needed
+    /**
+     * Generate a unique slug for the given monument
+     * @param monument  Monument to generate a slug for
+     */
+    public void generateSlug(Monument monument) {
+
+        List<Monument> duplicates;
+        try {
+            duplicates = getPossibleSlugDuplicates(monument);
+        } catch (HibernateException e) {
+            System.err.println("Hibernate Exception occurred while finding duplicate slugs");
+            throw e;
+        }
+
+        // Generate the slug from the Monument's title, city, and state
+        ArrayList<String> slugParts = new ArrayList<>(Arrays.asList(monument.getTitle(), monument.getCity(), monument.getState()));
+        // If these fields do not create a unique key, add on an additional identifier to make it unique
+        if (duplicates.size() > 0) slugParts.add(String.valueOf(duplicates.size()));
+
+        // Use the Slugify package to create a nice, readable slug from these fields
+        monument.setSlug(slugify.slugify(String.join(" ", slugParts)));
```

Comment:
Should we persist a Monument back to the database after we generate its Slug?

Reply:
Actually since you're just changing properties on the object reference, the changes get carried all the way back to hibernate and it completes the transaction with your changes made

Reply:
This is specifically because it's an event before insert and not after